src/common/datasource/src/util/io.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// modified from https://github.com/tokio-rs/tokio/blob/03912b9cf7141a5e27612a0b283d266637d2088f/tokio-util/src/io/sync_bridge.rs
```

Comment:
there are reasons that tokio prohibits in-runtime blocking. we'd better to comply it when whenever possible.

By the way, if we modify code from tokio, we should add their MIT license header in this file

Reply:
I removed it and used `tokio_util::io:: SyncIoBridge` instead, and I wrapped the blocking code with [tokio](https://docs.rs/tokio/latest/tokio/index.html)::[task](https://docs.rs/tokio/latest/tokio/task/index.html)::[spawn_blocking](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html#).

```rust
        tokio::task::spawn_blocking(move || {
            let mut reader = BufReader::new(SyncIoBridge::new(decoded));

            let iter = ValueIter::new(&mut reader, schema_infer_max_record);

            let schema = infer_json_schema_from_iterator(iter)
                .context(error::InferSchemaSnafu { path: &path })?;

            Ok(Arc::new(schema))
        })
        .await
        .context(error::JoinHandleSnafu)?
```

The document mentioned fits our use case, but I looked up the source code of hyper. However, it seems they have removed this approach.

> Another option is [SyncIoBridge](https://docs.rs/tokio-util/0.6/tokio_util/io/struct.SyncIoBridge.html) for cases where the synchronous context is operating on byte streams. For example, you might use an asynchronous HTTP client such as [hyper](https://docs.rs/hyper) to fetch data, but perform complex parsing of the payload body using a library written for synchronous I/O.

https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html